geoh5py/shared/conversion/base.py

```diff
@@ -0,0 +1,169 @@
+#
+#  geoh5py is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU Lesser General Public License for more details.
+#
+#  You should have received a copy of the GNU Lesser General Public License
+#  along with geoh5py.  If not, see <https://www.gnu.org/licenses/>.
+
+
+from __future__ import annotations
+
+from abc import ABC
+from typing import TYPE_CHECKING
+
+from ... import objects
+
+if TYPE_CHECKING:
+    from ...objects import BlockModel, Grid2D, ObjectBase, Octree, Points
+    from ...shared.entity import Entity
+    from ...workspace import Workspace
+
+entity_properties = ["name", "allow_rename", "allow_move", "allow_delete"]
+
+
+class ConversionBase(ABC):
+    _entity: ObjectBase
+
+    def __init__(self, entity: ObjectBase):
+        """
+        Converter class from an :obj:geoh5py.shared.entity.Entity to another.
+        :param entity: the entity to convert.
+        """
+        self._entity: Entity = entity
+        self._output = None
+        self._workspace_output: Workspace | None = None
+        self._name: str | None = None
+
+    def verify_kwargs(self, **kwargs):
+        """
+        Verify if the kwargs are valid.
+        :param kwargs: the kwargs to verify.
+        """
+        self._name = kwargs.get("name", self.entity.name)
+        self.change_workspace_parent(**kwargs)
+
+    def change_workspace_parent(self, **kwargs):
+        """
+        Define the workspace of the converter class if the workspace is defined in the kwargs;
+        else the workspace of the input entity is used.
+        """
+        if "workspace" in kwargs:
+            self._workspace_output = kwargs["workspace"]
+        else:
+            self._workspace_output = self._entity.workspace
+
+    def copy_properties(self):
+        """
+        Copy all the properties from the original entity to the new one.
+        """
+        if self._output is not None:
+            for property_ in entity_properties:
+                setattr(self._output, property_, getattr(self._entity, property_))
+        else:
+            raise ValueError("Output has not been created yet.")
+
+    @property
+    def workspace_output(self) -> Workspace:
+        """Workspace of the output object"""
+        if self._workspace_output is not None:
+            return self._workspace_output
+        raise ValueError(
+            "Workspace has not been defined yet,\
+        please run change_workspace_parent()."
+        )
+
+    @property
+    def entity(self) -> Entity:
+        """Input object"""
+        return self._entity
+
+    @property
+    def output(self):
+        """Output object"""
+        return self._output
+
+    @output.setter
+    def output(self, value):
+        """
+        Set the output object.
+        :param value: any values to be pass to output.
+        """
+        self._output = value
```

Comment:
I don't think we want to necessarily store one output. Remember that the converter can technically output to many type of objects. So we could call the same converter (for a given object) multiple times.